File: backend/routes/decks.py
from fastapi import APIRouter, HTTPException, Depends
import asyncpg
import uuid
from typing import Optional
from pydantic import BaseModel

router = APIRouter()


class DeckCreate(BaseModel):
    title: str
    description: Optional[str] = None
    privacy: str = "private"
    user_id: Optional[str] = None


class DeckUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    privacy: Optional[str] = None


async def get_db():
    from main import db_pool
    if db_pool is None:
        raise HTTPException(status_code=500, detail="Database not available")
    return db_pool


@router.get("/")
async def get_decks(user_id: Optional[str] = None, db=Depends(get_db)):
    async with db.acquire() as conn:
        if user_id:
            rows = await conn.fetch(
                "SELECT * FROM decks WHERE user_id = $1 OR privacy = $2 ORDER BY created_at DESC",
                user_id, "public"
            )
        else:
            rows = await conn.fetch(
                "SELECT * FROM decks WHERE privacy = $1 ORDER BY created_at DESC",
                "public"
            )
        result = []
        for row in rows:
            result.append({k: v for k, v in row.items()})
        return result


@router.get("/{deck_id}")
async def get_deck(deck_id: str, db=Depends(get_db)):
    async with db.acquire() as conn:
        row = await conn.fetchrow("SELECT * FROM decks WHERE id = $1", deck_id)
        if not row:
            raise HTTPException(status_code=404, detail="Deck not found")
        return dict(row)


@router.post("/")
async def create_deck(deck: DeckCreate, db=Depends(get_db)):
    if not deck.title:
        raise HTTPException(status_code=400, detail="Title is required")
    
    deck_id = str(uuid.uuid4())
    async with db.acquire() as conn:
        row = await conn.fetchrow(
            "INSERT INTO decks (id, user_id, title, description, privacy) VALUES ($1, $2, $3, $4, $5) RETURNING *",
            deck_id, deck.user_id, deck.title, deck.description, deck.privacy
        )
        return dict(row)


@router.put("/{deck_id}")
async def update_deck(deck_id: str, deck: DeckUpdate, db=Depends(get_db)):
    async with db.acquire() as conn:
        # Get existing deck
        existing = await conn.fetchrow("SELECT * FROM decks WHERE id = $1", deck_id)
        if not existing:
            raise HTTPException(status_code=404, detail="Deck not found")
        
        # Build update query
        updates = []
        values = []
        param_count = 1
        
        if deck.title is not None:
            updates.append(f"title = ${param_count}")
            values.append(deck.title)
            param_count += 1
        if deck.description is not None:
            updates.append(f"description = ${param_count}")
            values.append(deck.description)
            param_count += 1
        if deck.privacy is not None:
            updates.append(f"privacy = ${param_count}")
            values.append(deck.privacy)
            param_count += 1
        
        if not updates:
            return dict(existing)
        
        updates.append("updated_at = CURRENT_TIMESTAMP")
        values.append(deck_id)
        
        query = f"UPDATE decks SET {', '.join(updates)} WHERE id = ${param_count} RETURNING *"
        row = await conn.fetchrow(query, *values)
        return dict(row)


@router.delete("/{deck_id}")
async def delete_deck(deck_id: str, db=Depends(get_db)):
    async with db.acquire() as conn:
        row = await conn.fetchrow("DELETE FROM decks WHERE id = $1 RETURNING *", deck_id)
        if not row:
            raise HTTPException(status_code=404, detail="Deck not found")
        return {"message": "Deck deleted successfully"}

File: backend/src/routes/decks.js
const express = require('express');
const router = express.Router();
const { v4: uuidv4 } = require('uuid');

router.get('/', async (req, res) => {
  try {
    const { db } = req.app.locals;
    const { user_id } = req.query;
    
    let query = 'SELECT * FROM decks WHERE privacy = $1';
    const params = ['public'];
    
    if (user_id) {
      query = 'SELECT * FROM decks WHERE user_id = $1 OR privacy = $2 ORDER BY created_at DESC';
      params[0] = user_id;
      params[1] = 'public';
    }
    
    const result = await db.query(query, params);
    res.json(result.rows);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.get('/:id', async (req, res) => {
  try {
    const { db } = req.app.locals;
    const { id } = req.params;
    
    const result = await db.query('SELECT * FROM decks WHERE id = $1', [id]);
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Deck not found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post('/', async (req, res) => {
  try {
    const { db } = req.app.locals;
    const { title, description, privacy = 'private', user_id = null } = req.body;
    
    if (!title) {
      return res.status(400).json({ error: 'Title is required' });
    }
    
    const result = await db.query(
      'INSERT INTO decks (id, user_id, title, description, privacy) VALUES ($1, $2, $3, $4, $5) RETURNING *',
      [uuidv4(), user_id, title, description, privacy]
    );
    
    res.status(201).json(result.rows[0]);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.put('/:id', async (req, res) => {
  try {
    const { db } = req.app.locals;
    const { id } = req.params;
    const { title, description, privacy } = req.body;
    
    const updates = [];
    const values = [];
    let paramCount = 1;
    
    if (title !== undefined) {
      updates.push(`title = $${paramCount++}`);
      values.push(title);
    }
    if (description !== undefined) {
      updates.push(`description = $${paramCount++}`);
      values.push(description);
    }
    if (privacy !== undefined) {
      updates.push(`privacy = $${paramCount++}`);
      values.push(privacy);
    }
    
    updates.push(`updated_at = CURRENT_TIMESTAMP`);
    values.push(id);
    
    const query = `UPDATE decks SET ${updates.join(', ')} WHERE id = $${paramCount} RETURNING *`;
    const result = await db.query(query, values);
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Deck not found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.delete('/:id', async (req, res) => {
  try {
    const { db } = req.app.locals;
    const { id } = req.params;
    
    const result = await db.query('DELETE FROM decks WHERE id = $1 RETURNING *', [id]);
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Deck not found' });
    }
    
    res.json({ message: 'Deck deleted successfully' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

